File: python/src/equities_bench/harness/json_out.py
from __future__ import annotations

from pathlib import Path

from ..env import EnvInfo
from . import percentiles


class BenchResult:
    def __init__(self, name: str, ops: int, elapsed_ns: int, checksum: int, sorted_samples: list[int]) -> None:
        self.name = name
        self.ops = ops
        self.elapsed_ns = elapsed_ns
        self.throughput_ops_s = 0.0 if elapsed_ns == 0 else ops / (elapsed_ns / 1_000_000_000.0)
        self.checksum = str(checksum)
        self.min_ns = percentiles.min_v(sorted_samples)
        self.p50_ns = percentiles.pct(sorted_samples, 0.50)
        self.p90_ns = percentiles.pct(sorted_samples, 0.90)
        self.p99_ns = percentiles.pct(sorted_samples, 0.99)
        self.p999_ns = percentiles.pct(sorted_samples, 0.999)
        self.max_ns = percentiles.max_v(sorted_samples)
        self.stats: list[tuple[str, int]] = []

    def stat(self, k: str, v: int) -> "BenchResult":
        self.stats.append((k, v))
        return self


def _esc(s: str) -> str:
    return s.replace("\\", "\\\\").replace('"', '\\"')


def write(
    path: Path,
    env: EnvInfo,
    container_runtime: str,
    seed: int,
    ops: int,
    warmup: int,
    symbols: int,
    accounts: int,
    benches: list[BenchResult],
) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    lines = ["{\n"]
    def fs(k, v, comma=True, indent=2):
        lines.append(f'{" " * indent}"{k}": "{_esc(str(v))}"' + (",\n" if comma else "\n"))
    def fi(k, v, comma=True, indent=2):
        lines.append(f'{" " * indent}"{k}": {v}' + (",\n" if comma else "\n"))
    def ff(k, v, comma=True, indent=2):
        lines.append(f'{" " * indent}"{k}": {v:.6f}' + (",\n" if comma else "\n"))

    fi("schema_version", 1)
    fs("language", "python")
    fs("runtime", env.runtime)
    fs("os", env.os)
    fs("arch", env.arch)
    fi("cpus", env.cpus)
    fi("mem_bytes", env.mem_bytes)
    fs("os_pretty", env.os_pretty)
    fs("cpu_model", env.cpu_model)
    fs("container_runtime", container_runtime)
    fs("seed", str(seed))
    fi("ops", ops)
    fi("warmup", warmup)
    fi("symbols", symbols)
    fi("accounts", accounts)
    lines.append('  "benchmarks": [\n')
    for i, b in enumerate(benches):
        comma = i + 1 < len(benches)
        lines.append("    {\n")
        fs("name", b.name, True, 6)
        fi("ops", b.ops, True, 6)
        fi("elapsed_ns", b.elapsed_ns, True, 6)
        ff("throughput_ops_s", b.throughput_ops_s, True, 6)
        fs("checksum", b.checksum, True, 6)
        lines.append('      "latency_ns": {\n')
        fi("min", b.min_ns, True, 8)
        fi("p50", b.p50_ns, True, 8)
        fi("p90", b.p90_ns, True, 8)
        fi("p99", b.p99_ns, True, 8)
        fi("p999", b.p999_ns, True, 8)
        fi("max", b.max_ns, False, 8)
        lines.append("      },\n")
        lines.append('      "stats": {\n')
        for n, (k, v) in enumerate(b.stats):
            fi(k, v, n + 1 < len(b.stats), 8)
        lines.append("      }\n")
        lines.append("    }" + (",\n" if comma else "\n"))
    lines.append("  ]\n")
    lines.append("}\n")
    path.write_text("".join(lines), encoding="utf-8")

File: rust/src/json_out.rs
use std::fs;
use std::io::{self, Write};
use std::path::Path;

use crate::env_info::EnvInfo;

pub struct BenchResult {
    pub name: String,
    pub ops: u64,
    pub elapsed_ns: u64,
    pub throughput_ops_s: f64,
    pub checksum: String,
    pub min_ns: u64,
    pub p50_ns: u64,
    pub p90_ns: u64,
    pub p99_ns: u64,
    pub p999_ns: u64,
    pub max_ns: u64,
    pub stats: Vec<(String, i64)>,
}

impl BenchResult {
    pub fn new(
        name: &str,
        ops: u64,
        elapsed_ns: u64,
        checksum: u64,
        sorted: &[u64],
    ) -> Self {
        let throughput = if elapsed_ns == 0 {
            0.0
        } else {
            ops as f64 / (elapsed_ns as f64 / 1_000_000_000.0)
        };
        Self {
            name: name.to_string(),
            ops,
            elapsed_ns,
            throughput_ops_s: throughput,
            checksum: checksum.to_string(),
            min_ns: crate::percentiles::min(sorted),
            p50_ns: crate::percentiles::pct(sorted, 0.50),
            p90_ns: crate::percentiles::pct(sorted, 0.90),
            p99_ns: crate::percentiles::pct(sorted, 0.99),
            p999_ns: crate::percentiles::pct(sorted, 0.999),
            max_ns: crate::percentiles::max(sorted),
            stats: Vec::new(),
        }
    }

    pub fn stat(mut self, k: &str, v: i64) -> Self {
        self.stats.push((k.to_string(), v));
        self
    }
}

pub fn write(
    path: &Path,
    env: &EnvInfo,
    container_runtime: &str,
    seed: u64,
    ops: u64,
    warmup: u64,
    symbols: i32,
    accounts: i32,
    benches: &[BenchResult],
) -> io::Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let mut sb = String::new();
    sb.push_str("{\n");
    field_i(&mut sb, "schema_version", 1, true, 2);
    field_s(&mut sb, "language", "rust", true, 2);
    field_s(&mut sb, "runtime", &env.runtime, true, 2);
    field_s(&mut sb, "os", &env.os, true, 2);
    field_s(&mut sb, "arch", &env.arch, true, 2);
    field_i(&mut sb, "cpus", i64::from(env.cpus), true, 2);
    field_u(&mut sb, "mem_bytes", env.mem_bytes, true, 2);
    field_s(&mut sb, "os_pretty", &env.os_pretty, true, 2);
    field_s(&mut sb, "cpu_model", &env.cpu_model, true, 2);
    field_s(&mut sb, "container_runtime", container_runtime, true, 2);
    field_s(&mut sb, "seed", &seed.to_string(), true, 2);
    field_u(&mut sb, "ops", ops, true, 2);
    field_u(&mut sb, "warmup", warmup, true, 2);
    field_i(&mut sb, "symbols", i64::from(symbols), true, 2);
    field_i(&mut sb, "accounts", i64::from(accounts), true, 2);
    sb.push_str("  \"benchmarks\": [\n");
    for (i, b) in benches.iter().enumerate() {
        write_bench(&mut sb, b, i + 1 < benches.len());
    }
    sb.push_str("  ]\n");
    sb.push_str("}\n");
    let mut f = fs::File::create(path)?;
    f.write_all(sb.as_bytes())?;
    Ok(())
}

fn write_bench(sb: &mut String, b: &BenchResult, comma: bool) {
    sb.push_str("    {\n");
    field_s(sb, "name", &b.name, true, 6);
    field_u(sb, "ops", b.ops, true, 6);
    field_u(sb, "elapsed_ns", b.elapsed_ns, true, 6);
    field_f(sb, "throughput_ops_s", b.throughput_ops_s, true, 6);
    field_s(sb, "checksum", &b.checksum, true, 6);
    sb.push_str("      \"latency_ns\": {\n");
    field_u(sb, "min", b.min_ns, true, 8);
    field_u(sb, "p50", b.p50_ns, true, 8);
    field_u(sb, "p90", b.p90_ns, true, 8);
    field_u(sb, "p99", b.p99_ns, true, 8);
    field_u(sb, "p999", b.p999_ns, true, 8);
    field_u(sb, "max", b.max_ns, false, 8);
    sb.push_str("      },\n");
    sb.push_str("      \"stats\": {\n");
    for (n, (k, v)) in b.stats.iter().enumerate() {
        field_i(sb, k, *v, n + 1 < b.stats.len(), 8);
    }
    sb.push_str("      }\n");
    sb.push_str("    }");
    sb.push_str(if comma { ",\n" } else { "\n" });
}

fn field_s(sb: &mut String, k: &str, v: &str, comma: bool, indent: usize) {
    sb.push_str(&" ".repeat(indent));
    sb.push('"');
    sb.push_str(k);
    sb.push_str("\": \"");
    sb.push_str(&escape(v));
    sb.push('"');
    sb.push_str(if comma { ",\n" } else { "\n" });
}

fn field_i(sb: &mut String, k: &str, v: i64, comma: bool, indent: usize) {
    sb.push_str(&" ".repeat(indent));
    sb.push('"');
    sb.push_str(k);
    sb.push_str("\": ");
    sb.push_str(&v.to_string());
    sb.push_str(if comma { ",\n" } else { "\n" });
}

fn field_u(sb: &mut String, k: &str, v: u64, comma: bool, indent: usize) {
    sb.push_str(&" ".repeat(indent));
    sb.push('"');
    sb.push_str(k);
    sb.push_str("\": ");
    sb.push_str(&v.to_string());
    sb.push_str(if comma { ",\n" } else { "\n" });
}

fn field_f(sb: &mut String, k: &str, v: f64, comma: bool, indent: usize) {
    sb.push_str(&" ".repeat(indent));
    sb.push('"');
    sb.push_str(k);
    sb.push_str("\": ");
    sb.push_str(&format!("{v:.6}"));
    sb.push_str(if comma { ",\n" } else { "\n" });
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"")
}
